polkadot/zombienet_tests/functional/0007-dispute-freshly-finalized.toml

```diff
@@ -0,0 +1,45 @@
+[settings]
+timeout = 1000
+
+[relaychain.genesis.runtime.configuration.config]
+  max_validators_per_core = 2
+  needed_approvals = 3
+
+[relaychain]
+default_image = "{{ZOMBIENET_INTEGRATION_TEST_IMAGE}}"
+chain = "rococo-local"
+default_command = "polkadot"
+
+[relaychain.default_resources]
+limits = { memory = "4G", cpu = "2" }
+requests = { memory = "2G", cpu = "1" }
+
+  [[relaychain.node_groups]]
+  name = "honest"
+  count = 12
```

Comment:
I don't think this test requires 12 honest validators. Probably 3 is enough.

Reply:
On smaller networks disputes autoconfirm and are not properly filtered.

Reply:
with 1 malicious and 3 honest, the dispute will not be autoconfirmed, no? I assume you mean f+1 votes as confirmed

Reply:
That's what I initially thought as well and the first draft of the test was running using that setup. In that setting disputes will get insta confirmed in 100% of cases.

They way the system actually works is a bit different. Firstly, whenever you import a dispute statement you actually get a pair of 2 statements (1 valid, 1 invalid). So that already puts you at 2 total votes sent to the honest nodes. In our case the invalid statement comes from the malus node and it is accompanied by a backing statement observed previously.

Additionally the honest node to which malus is sending the statement usually already has 2 backing statements in his db. So assuming the malus was NOT the backer for this candidate the honest node will have 3 different statements: 2x valid statements from honest backing + 1x invalid statement from malus. 

3 statements will lead to a dispute confirmation in networks 10 or smaller in size.
byzantine_th = 10-1//3 -> 3
byzantine_th +1 is needed for a confirmation

---

Played around with it a bit more and what we can do is lower the `max_validators_per_core` to 1 and then we reduce the number of backing votes in the honest node db. This allows us to reduce the size of the network to 7 in total. Unfortunately 4 will still not work as we'll get a minimum of 2 votes which is higher than our byzantine assumption.

Everything said above I tested in practice on all cases and verified the votes through logs so it's not just speculation.

Hope that clears it up and convinces you :smile: @ordian

Reply:
Thanks for addressing the feedback. Yep, that's convincing :) Although, can it be 5 honest, 6 in total? 2 out of 6 is < f+1.

Reply:
> Although, can it be 5 honest, 6 in total? 2 out of 6 is < f+1.

Short answer is no. I've tested this before and it autoconfirms on 6. 

Reasoning for that lies in the way we calculate `byzantine_threshold`. In the case of `network_size=6` -> `f=1` because `f=(6-1)//3=1` it's fine because `6=3f+3`. In general `f` is always such that `network_size = 3f + r`, where `r ∈ {1,2,3} and f ∈ ℕ`.

Going to a `network_size=7` gives us `f=(7-1)//3 = 2`. We generally need it always to be 1 higher than a multiple of 3 for the minimal size. `f=2` means that we need 3 votes to confirm a dispute, but on `max_validators_per_core=1` we only get 1 backer + 1 from malus.